examples/microGPT.py

```diff
@@ -0,0 +1,328 @@
+        rank_zero_info("data has %d characters, %d unique." % (data_size, vocab_size))
+
+        self.stoi = {ch: i for i, ch in enumerate(chars)}
+        self.itos = {i: ch for i, ch in enumerate(chars)}
+        self.block_size = block_size
+        self.vocab_size = vocab_size
+        self.data = data
+
+    def __len__(self):
+        return len(self.data) - self.block_size
+
+    def __getitem__(self, i):
+        chunk = self.data[i : i + self.block_size + 1]
+        dix = [self.stoi[s] for s in chunk]
+
+        # src and target are off by one, we want the model to predict the next word
+        x = torch.tensor(dix[:-1], dtype=torch.long)
+        y = torch.tensor(dix[1:], dtype=torch.long)
+        return x, y
+
+    def to_tokens(self, message, device):
+        return torch.tensor([self.stoi[s] for s in message], dtype=torch.long)[
+            None, ...
+        ].to(device)
+
+    def from_tokens(self, tokens):
+        return "".join([self.itos[int(i)] for i in tokens])
+
+
+@torch.no_grad()
+def sample(model, x, steps, temperature=1.0, sample=False, top_k=None):
+    """
+    take a conditioning sequence of indices in x (of shape (b,t)) and predict the next token in
+    the sequence, feeding the predictions back into the model each time. Clearly the sampling
+    has quadratic complexity unlike an RNN that is only linear, and has a finite context window
+    of block_size, unlike an RNN that has an infinite context window.
+    """
+    block_size = model.get_block_size()
+    model.eval()
+
+    # CREDITS: https://github.com/karpathy/minGPT/blob/master/mingpt/utils.py
+    def top_k_logits(logits, k):
+        v, _ = torch.topk(logits, k)
+        out = logits.clone()
+        out[out < v[:, [-1]]] = -float("Inf")
+        return out
+
+    for _ in range(steps):
+        # FIXME: Probably a bug here in the initial context cropping
+        x_cond = (
+            x if x.size(1) <= block_size else x[:, -block_size:]
+        )  # crop context if needed
+        logits = model(x_cond)
+
+        # pluck the logits at the final step and scale by temperature
+        logits = logits[:, -1, :] / temperature
+
+        # optionally crop probabilities to only the top k options
+        if top_k is not None:
+            logits = top_k_logits(logits, top_k)
+
+        # apply softmax to convert to probabilities
+        probs = F.softmax(logits, dim=-1)
+
+        # sample from the distribution or take the most likely
+        if sample:
+            ix = torch.multinomial(probs, num_samples=1)
+        else:
+            _, ix = torch.topk(probs, k=1, dim=-1)
+
+        # append to the sequence and continue
+        x = torch.cat((x, ix), dim=1)
+
+    return x
+
+
+if __name__ == "__main__":
+    seed_everything(42)
+    N_LAYERS = 8
+    N_HEADS = 8
+    EMB = 512
+    LR = 6e-4
+    BLOCK = 128
+    REF_BATCH = 512
+    BATCH = 256
+    WORKERS = 4
+    EPOCHS = 1
+    WARMUP = 20
```

Comment:
do they differ from defaults used in GPT init? if not let's use only one...

Reply:
Yes, great point, thank you ! Will update tomorrow